File: web/src/api/endpoints/stats.ts
import { useQuery } from '@tanstack/react-query';
import { apiClient } from '../client';
import { formatCount, formatCurrencyCosts, formatTime, type CostCurrencyMetrics } from '@/lib/utils';

/**
 * 统计数据
 */
export interface StatsMetrics {
    input_token: number;
    output_token: number;
    input_cost: number;
    output_cost: number;
    input_cost_by_currency?: Record<string, CostCurrencyMetrics>;
    output_cost_by_currency?: Record<string, CostCurrencyMetrics>;
    total_cost_by_currency?: Record<string, CostCurrencyMetrics>;
    wait_time: number;
    request_success: number;
    request_failed: number;
}

export interface StatsMetricsFormatted {
    input_token: ReturnType<typeof formatCount>;
    output_token: ReturnType<typeof formatCount>;
    input_cost: ReturnType<typeof formatCurrencyCosts>;
    output_cost: ReturnType<typeof formatCurrencyCosts>;
    wait_time: ReturnType<typeof formatTime>;
    request_success: ReturnType<typeof formatCount>;
    request_failed: ReturnType<typeof formatCount>;

    request_count: ReturnType<typeof formatCount>;
    total_token: ReturnType<typeof formatCount>;
    total_cost: ReturnType<typeof formatCurrencyCosts>;
}

export interface StatsChannel extends StatsMetrics {
    channel_id: number;
}

export interface StatsChannelKey extends StatsMetrics {
    channel_key_id: number;
}

export interface StatsDaily extends StatsMetrics {
    date: string;
}
export interface StatsDailyFormatted extends StatsMetricsFormatted {
    date: string;
}

export interface StatsTotal extends StatsMetrics {
    id: number;
}
export type StatsTotalFormatted = StatsMetricsFormatted;

export interface StatsHourly extends StatsMetrics {
    hour: number;
    date: string;
}
export interface StatsHourlyFormatted extends StatsMetricsFormatted {
    hour: number;
    date: string;
}
/**
 * API Key 统计数据
 */
export interface StatsAPIKey extends StatsMetrics {
    api_key_id: number;
}

export interface StatsAPIKeyFormatted extends StatsMetricsFormatted {
    api_key_id: number;
}
/**
 * 获取今日统计数据 Hook
 */
export function useStatsToday() {
    return useQuery({
        queryKey: ['stats', 'today'],
        queryFn: async () => {
            return apiClient.get<StatsDaily>('/api/v1/stats/today');
        },
        refetchInterval: 30000,
        refetchOnMount: 'always',
    });
}

/**
 * 获取每日统计数据 Hook
 */
export function useStatsDaily() {
    return useQuery({
        queryKey: ['stats', 'daily'],
        queryFn: async () => {
            return apiClient.get<StatsDaily[]>('/api/v1/stats/daily');
        },
        select: (data) => data.map((item): StatsDailyFormatted => ({
            input_token: formatCount(item.input_token),
            output_token: formatCount(item.output_token),
            total_token: formatCount(item.input_token + item.output_token),
            input_cost: formatCurrencyCosts(item.input_cost_by_currency, item.input_cost),
            output_cost: formatCurrencyCosts(item.output_cost_by_currency, item.output_cost),
            total_cost: formatCurrencyCosts(item.total_cost_by_currency, item.input_cost + item.output_cost),
            wait_time: formatTime(item.wait_time),
            request_success: formatCount(item.request_success),
            request_failed: formatCount(item.request_failed),
            request_count: formatCount(item.request_success + item.request_failed),
            date: item.date,
        })),
        refetchInterval: 3600000, // 1 小时
        refetchOnMount: 'always',
    });
}
/**
 * 获取总统计数据 Hook
 */
export function useStatsHourly() {
    return useQuery({
        queryKey: ['stats', 'hourly'],
        queryFn: async () => {
            return apiClient.get<StatsHourly[]>('/api/v1/stats/hourly');
        },
        select: (data) => data.map((item): StatsHourlyFormatted => ({
            hour: item.hour,
            date: item.date,
            input_token: formatCount(item.input_token),
            output_token: formatCount(item.output_token),
            total_token: formatCount(item.input_token + item.output_token),
            input_cost: formatCurrencyCosts(item.input_cost_by_currency, item.input_cost),
            output_cost: formatCurrencyCosts(item.output_cost_by_currency, item.output_cost),
            total_cost: formatCurrencyCosts(item.total_cost_by_currency, item.input_cost + item.output_cost),
            wait_time: formatTime(item.wait_time),
            request_success: formatCount(item.request_success),
            request_failed: formatCount(item.request_failed),
            request_count: formatCount(item.request_success + item.request_failed),
        })),
        refetchInterval: 10000,// 10 秒
        refetchOnMount: 'always',
    });
}

export function useStatsTotal() {
    return useQuery({
        queryKey: ['stats', 'total'],
        queryFn: async () => {
            return apiClient.get<StatsTotal>('/api/v1/stats/total');
        },
        select: (data) => ({
            input_token: formatCount(data.input_token),
            output_token: formatCount(data.output_token),
            total_token: formatCount(data.input_token + data.output_token),
            input_cost: formatCurrencyCosts(data.input_cost_by_currency, data.input_cost),
            output_cost: formatCurrencyCosts(data.output_cost_by_currency, data.output_cost),
            total_cost: formatCurrencyCosts(data.total_cost_by_currency, data.input_cost + data.output_cost),
            wait_time: formatTime(data.wait_time),
            request_success: formatCount(data.request_success),
            request_failed: formatCount(data.request_failed),
            request_count: formatCount(data.request_success + data.request_failed),
        }),
        refetchInterval: 10000,// 10 秒
        refetchOnMount: 'always',
    });
}



/**
 * 获取 API Key 统计数据列表 Hook
 */
export function useStatsAPIKey() {
    return useQuery({
        queryKey: ['stats', 'apikey'],
        queryFn: async () => {
            return apiClient.get<StatsAPIKey[]>('/api/v1/stats/apikey');
        },
        select: (data) => data.map((item): StatsAPIKeyFormatted => ({
            api_key_id: item.api_key_id,
            input_token: formatCount(item.input_token),
            output_token: formatCount(item.output_token),
            total_token: formatCount(item.input_token + item.output_token),
            input_cost: formatCurrencyCosts(item.input_cost_by_currency, item.input_cost),
            output_cost: formatCurrencyCosts(item.output_cost_by_currency, item.output_cost),
            total_cost: formatCurrencyCosts(item.total_cost_by_currency, item.input_cost + item.output_cost),
            wait_time: formatTime(item.wait_time),
            request_success: formatCount(item.request_success),
            request_failed: formatCount(item.request_failed),
            request_count: formatCount(item.request_success + item.request_failed),
        })),
        refetchInterval: 30000,
        refetchOnMount: 'always',
    });
}

File: internal/model/stats.go
package model

type StatsMetrics struct {
	InputToken           int64                          `json:"input_token" gorm:"bigint"`
	OutputToken          int64                          `json:"output_token" gorm:"bigint"`
	InputCost            float64                        `json:"input_cost" gorm:"type:real"`
	OutputCost           float64                        `json:"output_cost" gorm:"type:real"`
	InputCostByCurrency  map[string]CostCurrencyMetrics `json:"input_cost_by_currency,omitempty" gorm:"serializer:json"`
	OutputCostByCurrency map[string]CostCurrencyMetrics `json:"output_cost_by_currency,omitempty" gorm:"serializer:json"`
	TotalCostByCurrency  map[string]CostCurrencyMetrics `json:"total_cost_by_currency,omitempty" gorm:"serializer:json"`
	WaitTime             int64                          `json:"wait_time" gorm:"bigint"`
	RequestSuccess       int64                          `json:"request_success" gorm:"bigint"`
	RequestFailed        int64                          `json:"request_failed" gorm:"bigint"`
}

type StatsTotal struct {
	ID int `gorm:"primaryKey"`
	StatsMetrics
}

type StatsHourly struct {
	Hour int    `json:"hour" gorm:"primaryKey"`
	Date string `json:"date" gorm:"not null"` // 记录最后更新日期，格式：20060102
	StatsMetrics
}

type StatsDaily struct {
	Date string `json:"date" gorm:"primaryKey"`
	StatsMetrics
}

type StatsModel struct {
	ID        int    `json:"id" gorm:"primaryKey"`
	Name      string `json:"name" gorm:"not null;uniqueIndex:idx_stats_model_name_channel"`
	ChannelID int    `json:"channel_id" gorm:"not null;uniqueIndex:idx_stats_model_name_channel"`
	StatsMetrics
}

type StatsChannel struct {
	ChannelID int `json:"channel_id" gorm:"primaryKey"`
	StatsMetrics
}

type StatsChannelKey struct {
	ChannelKeyID int `json:"channel_key_id" gorm:"primaryKey"`
	StatsMetrics
}

type StatsAPIKey struct {
	APIKeyID int `json:"api_key_id" gorm:"primaryKey"`
	StatsMetrics
}

// Add aggregates another StatsMetrics into the current one.
func (s *StatsMetrics) Add(delta StatsMetrics) {
	s.InputToken += delta.InputToken
	s.OutputToken += delta.OutputToken
	s.InputCost += delta.InputCost
	s.OutputCost += delta.OutputCost
	mergeCostCurrencyMetrics(&s.InputCostByCurrency, delta.InputCostByCurrency)
	mergeCostCurrencyMetrics(&s.OutputCostByCurrency, delta.OutputCostByCurrency)
	mergeCostCurrencyMetrics(&s.TotalCostByCurrency, delta.TotalCostByCurrency)
	s.WaitTime += delta.WaitTime
	s.RequestSuccess += delta.RequestSuccess
	s.RequestFailed += delta.RequestFailed
}

func (s *StatsMetrics) AddCurrencyCosts(info PricingInfo, inputCost, outputCost float64) {
	key := CostCurrencyKey(info.Currency)
	value := CostCurrencyMetrics{
		Currency:       key,
		CurrencySymbol: info.CurrencySymbol,
		InputCost:      inputCost,
		OutputCost:     outputCost,
		TotalCost:      inputCost + outputCost,
	}
	addCostCurrencyMetric(&s.InputCostByCurrency, key, CostCurrencyMetrics{
		Currency:       key,
		CurrencySymbol: info.CurrencySymbol,
		InputCost:      inputCost,
		TotalCost:      inputCost,
	})
	addCostCurrencyMetric(&s.OutputCostByCurrency, key, CostCurrencyMetrics{
		Currency:       key,
		CurrencySymbol: info.CurrencySymbol,
		OutputCost:     outputCost,
		TotalCost:      outputCost,
	})
	addCostCurrencyMetric(&s.TotalCostByCurrency, key, value)
}

func mergeCostCurrencyMetrics(dst *map[string]CostCurrencyMetrics, src map[string]CostCurrencyMetrics) {
	for key, value := range src {
		addCostCurrencyMetric(dst, key, value)
	}
}

func addCostCurrencyMetric(dst *map[string]CostCurrencyMetrics, key string, value CostCurrencyMetrics) {
	if key == "" {
		key = CostCurrencyKey(value.Currency)
	}
	if *dst == nil {
		*dst = make(map[string]CostCurrencyMetrics)
	}
	current := (*dst)[key]
	if current.Currency == "" {
		current.Currency = key
	}
	if current.CurrencySymbol == "" {
		current.CurrencySymbol = value.CurrencySymbol
	}
	current.InputCost += value.InputCost
	current.OutputCost += value.OutputCost
	current.TotalCost += value.TotalCost
	(*dst)[key] = current
}
